src/workerd/api/node/zlib-util.h

```diff
@@ -1,57 +1,261 @@
 // Copyright (c) 2017-2022 Cloudflare, Inc.
 // Licensed under the Apache 2.0 license found in the LICENSE file or at:
 //     https://opensource.org/licenses/Apache-2.0
-
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/exception.h>
 
+#include "zlib.h"
 #include <kj/array.h>
 #include <kj/compat/brotli.h>
-#include "zlib.h"
+#include <kj/mutex.h>
+#include <kj/vector.h>
 
-#include <limits>
-#include <cstdint>
+#include <cstdlib>
 
 namespace workerd::api::node {
 
+#ifndef ZLIB_ERROR_CODES
+#define ZLIB_ERROR_CODES(V)      \
+  V(Z_OK)                        \
+  V(Z_STREAM_END)                \
+  V(Z_NEED_DICT)                 \
+  V(Z_ERRNO)                     \
+  V(Z_STREAM_ERROR)              \
+  V(Z_DATA_ERROR)                \
+  V(Z_MEM_ERROR)                 \
+  V(Z_BUF_ERROR)                 \
+  V(Z_VERSION_ERROR)             \
+
+inline const char* ZlibStrerror(int err) {
+#define V(code) if (err == code) return #code;
+  ZLIB_ERROR_CODES(V)
+#undef V
+  return "Z_UNKNOWN_ERROR";
+}
+#endif  // ZLIB_ERROR_CODES
+
+// Certain zlib constants are defined by Node.js itself
+static constexpr auto Z_MIN_CHUNK = 64;
+static constexpr auto Z_MAX_CHUNK = 128 * 1024;
+static constexpr auto Z_DEFAULT_CHUNK = 16 * 1024;
+static constexpr auto Z_MIN_MEMLEVEL = 1;
+
+static constexpr auto Z_MAX_MEMLEVEL = 9;
+static constexpr auto Z_DEFAULT_MEMLEVEL = 8;
+static constexpr auto Z_MIN_LEVEL = -1;
+static constexpr auto Z_MAX_LEVEL = 9;
+static constexpr auto Z_DEFAULT_LEVEL = Z_DEFAULT_COMPRESSION;
+static constexpr auto Z_MIN_WINDOWBITS = 8;
+static constexpr auto Z_MAX_WINDOWBITS = 15;
+static constexpr auto Z_DEFAULT_WINDOWBITS = 15;
+
+static constexpr uint8_t GZIP_HEADER_ID1 = 0x1f;
+static constexpr uint8_t GZIP_HEADER_ID2 = 0x8b;
+
+using ZlibModeValue = uint8_t;
+enum class ZlibMode : ZlibModeValue {
+  NONE,
+  DEFLATE,
+  INFLATE,
+  GZIP,
+  GUNZIP,
+  DEFLATERAW,
+  INFLATERAW,
+  UNZIP,
+  BROTLI_DECODE,
+  BROTLI_ENCODE
+};
+
+struct CompressionError {
+  CompressionError(kj::StringPtr _message, kj::StringPtr _code, int _err)
```

Comment:
I don't personally like variable names starting with `_` but if it's a convention elsewhere in our codebase, then ignore me.

Reply:
This is a pretty common Node.js convention, but I'm pretty sure @jasnell uses it a lot as well. How would you initialize CompressionError where there is a `message` attribute? How would you name this variable int that scenario?

Reply:
Same name for both is what I'm used to, but if we're doing underscores I'll change.

Reply:
For this code that is largely adopted from Node.js I prefer that we maintain the Node.js conventions in order to make porting (and more importantly updating later) easier